Fix updating slottables when slots are removed from a shadowtree

TimvdLippe opened the pull request:
We never cleared the containing shadowRoot when unbinding
a node, which meant that slots wouldn't update their
slottables.

jdm commented on components/script/dom/node.rs:
```diff
@@ -406,6 +406,13 @@ impl Node {
+        // the containing root to clear up data.
+        for node in root.traverse_preorder(ShadowIncluding::No) {
+            node.set_containing_shadow_root(None);
+        }
```
One other option would be to add a containing_root field to the unbind context argument, so that the elements can still read the required information.

TimvdLippe replied:
I think I have a potential solution in mind. We can change the structure of the second loop to mirror the structure of the first loop. E.g. we manually traverse shadow boundaries. Then I can reset the containing shadow root after calling unbind in the second loop, rather than the first loop. But without introducing a third loop, because of performance.

Won't have time until tomorrow evening to test my theory, but I am pretty sure that will work.

simonwuelker replied:
If this works can you submit the fix in a standalone PR? Because to correct the condition `assign_slottables_in_a_tree` I believe I need https://github.com/servo/servo/issues/40242 which in turn needs this fix.

TimvdLippe replied:
Yes will attempt that this evening and let you know

TimvdLippe committed:
Fix updating slottables when slots are removed from a shadowtree

We never cleared the containing shadowRoot when unbinding
a node, which meant that slots wouldn't update their
slottables.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe commented on components/script/dom/node.rs:
```diff
@@ -386,23 +386,52 @@ impl Node {
+        // Use a macro instead of a function/closure as this is a hot path.
+        // Since both the initial traversal in light dom and the inner traversal
+        // in shadow DOM share the same code, we define a macro to prevent omissions.
+        macro_rules! cleanup_node(
```
This cannot be a function as we use `is_parent_connected` and such. Therefore, it would need to become a closure and afaik closures cannot be inlined, can they? We could inject all arguments as well, but wouldn't a macro then be more straightforward?

TimvdLippe replied:
Done. Wasn't able to add `#[inline]` as that leads to

```
error[E0658]: attributes on expressions are experimental
   --> components/script/dom/node.rs:392:28
    |
392 |         let cleanup_node = #[inline] |node: &Node| {
    |                            ^^^^^^^^^
    |
    = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
```

simonwuelker replied:
yeah you can't add annotations to closures.

simonwuelker reviewed (approved):
Thanks for the fix! Nice comments as well.
